### PR TITLE
C#: Enable nullability in Extraction.CSharp

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR is the fourth in multiple steps to enable nullability analysis and remove nullability warnings. It's based on #5189. This PR enables nullability analysis on the Extraction.CSharp project. I added some `#nullable disable warnings` directives where I couldn't get around some warnings. The biggest such section is in the `Analyser.cs`, where we have a two phase initialization with the `InitializeStandalone`/`EndInitialize`, and therefore we have a lot of nullable fields. I think it's not worth the effort to try to fix these. (It seems quite cumbersome to do it without changing our log messages, which might be considered a breaking change.)

The last commit is independent of nullability analysis. It combines the standalone and normal analysis into a single method. However, it needs quite a few `Func/Action` parameters, so in the end it didn't help too much with readability. Nevertheless, I think it would make sense to combine these two paths.

~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/887/) [Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/889/)~~ [Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/893/)

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction.CSharp/Analyser.cs [hunk 1]
```diff
@@ -10,22 +10,24 @@
 using Semmle.Util.Logging;
 using Semmle.Util;
 
+#nullable disable warnings
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Analyser.cs:13]: Todo: This file needs to be revisited.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/CommentLine.cs [hunk 2]
```diff
@@ -1,25 +1,25 @@
-using Semmle.Extraction.CommentProcessing;
 using Semmle.Extraction.Entities;
 using System.IO;
+using System;
 
 namespace Semmle.Extraction.CSharp.Entities
 {
-    internal class CommentLine : CachedEntity<(Microsoft.CodeAnalysis.Location, string)>, ICommentLine
+    internal class CommentLine : CachedEntity<Tuple<Microsoft.CodeAnalysis.Location, string>>
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/CommentLine.cs:7]: There's a non-cosmetic change here: `ValueTuple -> Tuple`. This way the `class` generic type constraint could be applied on the type parameter of the base class.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/CachedEntity.cs [hunk 3]
```diff
@@ -1,11 +1,12 @@
 namespace Semmle.Extraction.CSharp.Entities
 {
-    internal abstract class CachedEntity<T> : Extraction.CachedEntity<T>
+    internal abstract class CachedEntity<T> : Extraction.CachedEntity<T> where T : class
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/CachedEntity.cs:3]: Would it make sense to replace `class` with `notnull`?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/CommentLine.cs (hunk 2 above)
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/CommentLine.cs:7]: I don't think this change is needed if you replace the `class` constraint with `notnull`.

File: csharp/extractor/Semmle.Extraction.CSharp/Extractor/NonStandaloneAnalyser.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+using System.IO;
+using System.Linq;
+using System.Collections.Generic;
+using Semmle.Util.Logging;
+using Semmle.Util;
+
+
+namespace Semmle.Extraction.CSharp
+{
+    public class NonStandaloneAnalyser : Analyser, IDisposable
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Extractor/NonStandaloneAnalyser.cs:13]: Perhaps replace `NonStandalone` with `Tracing` throughout.

File: csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs [hunk 5]
```diff
@@ -321,73 +248,175 @@ private static IEnumerable<Action> ReadSyntaxTrees(IEnumerable<string> sources,
...
+                    {
+                        progressMonitor.MissingType(type);
+                    }
 
-                foreach (var type in analyser.MissingTypes)
+                    progressMonitor.MissingSummary(analyser.MissingTypes.Count(), analyser.MissingNamespaces.Count());
+                });
+        }
+
+        private static ExitCode AnalyseNonStandalone(
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs:372]: Should this be in `NonStandaloneExtractor.cs`?

File: csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs [hunk 6]
```diff
@@ -321,73 +248,175 @@ private static IEnumerable<Action> ReadSyntaxTrees(IEnumerable<string> sources,
...
+
+            logPerformance(performance);
+            analyser.Logger.Log(Severity.Info, "  Extraction took {0}", sw.Elapsed);
+
+            postProcess();
+
+            return analyser.TotalErrors == 0 ? ExitCode.Ok : ExitCode.Errors;
+        }
+
+        private static void AnalyseStandalone(
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs:341]: Should this be in `StandaloneExtractor.cs`?

**Review by hvitved — COMMENTED** [review]
Looks good to me; a few minor suggestions.

File: csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs (hunk 5 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs:372]: I didn't want to change even more things. I think the best would be to get rid of the `public static class Extractor` altogether and move its bits maybe to `StandaloneAnalyser` and `TracingAnalyser`.

**Review by hvitved — APPROVED** [review]

**Pull request merged**